### PR TITLE
Encodes blazingly fastly

### PR DESCRIPTION
**Pull request by andyquinterom**
Gets very close to geohashTools. Sadly most of the time is spent asking R for memory. We should try to create our own container type. Maybe we could use some sort of vector type. In a 64 bit number we can fit up to 12 characters. In two 64 bit values we could bit 24 (something like how date time is managed or bit64)

Allocating characters in R is really bad

### CODE REVIEW

**Comment by andyquinterom** [comment]
```
n <- 1e7
x <- runif(n, -180, 180)
y <- runif(n, -90, 90)

bench::mark(
  geohash = geohash::encode(x, y, 8L),
  geohash_par = geohash::encode_par(x, y, 8L),
  geohashTools = geohashTools::gh_encode(y, x, 8L)
)
# A tibble: 3 × 13
  expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr>   <bch:> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 geohash       2.84s  2.84s     0.352   204.1MB    0         1     0      2.84s
2 geohash_par   2.54s  2.54s     0.393    76.3MB    0.393     1     1      2.54s
3 geohashTools  2.45s  2.45s     0.408    76.3MB    0         1     0      2.45s
# ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
```

Minor stuff is needed to get closer. Parallelization barely helps due to the nature of R's memory model for strings.

**Comment by andyquinterom** [comment]
image

File: src/rust/src/encode.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+/// @export
+fn encode(longitude: &[f64], latitude: &[f64], length: i32) -> Robj {
+    let n = latitude.len();
+    let k = length as usize;
+
+    if k > MAX_K {
+        panic!("Precsion limit reached");
+    }
+
+    let map: [[u8; 8]; 4] = [*b"0145hjnp", *b"2367kmqr", *b"89destwx", *b"bcfguvyz"];
```
Comment (JosiahParry) [inline src/rust/src/encode.rs:76]: we can make this a const which should reduce compute time as well

**Comment by JosiahParry** [comment]
i pushed a small change which chunks the parallel iterator making this implementation faster than geohashTools

image

The remaining issue is the stack warning which might be due to the protect call `Warning: stack imbalance in '<-', 2 then 4` I'm worried that might lead to a memory leak.

**Comment by andyquinterom** [comment]
Let me check if I made a mistake that's easy to miss

**Comment by andyquinterom** [comment]
> i pushed a small change which chunks the parallel iterator making this implementation faster than geohashTools
> 
> 
> 
> image
> 
> 
> 
> The remaining issue is the stack warning which might be due to the protect call `Warning: stack imbalance in '<-', 2 then 4` I'm worried that might lead to a memory leak.

image

I'm doing a double protect :/

**Comment by andyquinterom** [comment]
Once I got so free time I'll fix it

**Comment by JosiahParry** [comment]
The finish review thing kills me! 

I've pushed to your PR and resolved it!

**Pull request merged**